api/src/main/java/com/datatorrent/api/AffinityRule.java

```diff
@@ -0,0 +1,211 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.api;
+
+import java.io.Serializable;
+import java.util.List;
+
+import com.datatorrent.api.DAG.Locality;
+
+/**
+ * Affinity rule specifies constraints for physical deployment of operator
+ * containers. There are two types of rules that can be specified: Affinity and
+ * Anti-affinity. Each rule contains list of operators or pair of 2 operators or
+ * a regex that should match at least 2 operators. Based on the type of rule,
+ * affinity or anti-affinity, the operators will be deployed together or away
+ * from each other. The locality indicates the level at which the rule should be
+ * applied. E.g. CONTAINER_LOCAL affinity would indicate operators Should be
+ * allocated within same container NODE_LOCAL anti-affinity indicates that the
+ * operators should not be allocated on the same node. The rule can be either
+ * strict or relaxed.
+ *
+ */
+public class AffinityRule implements Serializable
+{
+  @Override
+  public String toString()
+  {
+    return "AffinityRule {operatorsList=" + operatorsList + ", operatorRegex=" + operatorRegex + ", operators="
+        + operators + ", locality=" + locality + ", type=" + type + ", relaxLocality=" + relaxLocality + "}";
+  }
+
+  private static final long serialVersionUID = 107131504929875386L;
+
+  /**
+   * Pair of operator names to specify affinity rule
+   * The order of operators is not considered in this class
+   * i.e. OperatorPair("O1", "O2") is equal to OperatorPair("O2", "O1")
+   */
+  public static class OperatorPair implements Serializable
```

Comment:
Collection support is for convenience. Instead of saying anti affinity between operator pairs lets say A & B , B & C, with collection we can directly set anti-affinity between collection A, B, C. 
Internally, everything is converted into operator pairs, as this is easier to validate against stream locality and other affinity rules.

The OperatorPair class was created because Pair class inside apex-common is not available under api.